### ISSUE TITLE
Analyzer high memory consumption, "Too many files open" errors

### ISSUE THREAD

**Issue by johnbland-wf** [Dart-Code/Dart-Code#317]
You can see the `DanTup.dart` file has two separate processes running at the same time and is chewing up a lot of memory. Not long after running the below `ps aux` I checked Activity Monitor and the process was chewing through 1.9 GB.

There seems to be a memory leak. I believe this causes the problem of "too many files open" in Terminal after using this for a day or so.

```
johnbland        90176   0.0  2.3  2937680 381884   ??  S    10:45AM   0:19.31 /usr/local/Cellar/dart/1.23.0/libexec/bin/dart /usr/local/Cellar/dart/1.23.0/libexec/bin/snapshots/analysis_server.dart.snapshot --client-id=DanTup.dart-code --client-version=1.4.1

johnbland        90177   0.0  3.7  3781508 626064   ??  S    10:45AM   2:12.88 /usr/local/Cellar/dart/1.23.0/libexec/bin/dart /usr/local/Cellar/dart/1.23.0/libexec/bin/snapshots/analysis_server.dart.snapshot --client-id=DanTup.dart-code --client-version=1.4.1
```

**Comment by DanTup** [comment]
@johnbland-wf I presume you only had one Code project open at this time (so two processes is unexpected)? Did you have anything like a Code crash at all in this period? (is 10:45AM the process start time? Definitely should not be spawning two of these at the same time!)

It seems like there may be two issues here:

1. Dart Code has spawned two copies of the analyzer when it shouldn't
2. The analyzer is consuming too much memory

I can investigate (2), but I think (1) might be something for https://github.com/dart-lang/sdk

> I believe this causes the problem of "too many files open" in Terminal 

I've never seen this error before but from a quick search it seems like this is an OS-level limit on how many files a process can open. There's a command [in this answer](https://superuser.com/a/1180084) that might let you see which process it is which might help us identify whether it's inside the Dart VM (analyzer) or something in Dart Code. Are you able to try this next time you see it?

Thanks!

**Comment by johnbland-wf** [comment]
Apologies...there were two Code windows opened. I was mostly focused on the memory issue here.

**Comment by Skylled** [comment]
Would this be a Dart-Code problem or an sdk problem? My Dart analyzer process is sitting at around 400mb of RAM usage after 30 minutes of VS Code being open. Does Atom's dartlang plugin have this issue?

**Comment by DanTup** [comment]
> Apologies...there were two Code windows opened.

Thanks for confirming!

I suspect this will be an analyzer issue and need to go to dart-lang/sdk. I'm not very familiar with Observatory, but I believe it could be used to examine the memory by attaching it to the Analyzer process (unfortunately you'd need to relaunch the analyzer with a flag to enable this and then reproduce it). If you think you could do this, I can test it out and type up instructions for it shortly?

**Comment by johnbland-wf** [comment]
@Skylled, I never saw this issue with Atom's; used it for well over a year. I am at about 400 mb right now as well after restarting the process and VS Code.

@DanTup, the analyzer is kicked off in VS Code so I assume I'd need to add a setting to change it? Either way, let me know and I can do that.

**Comment by johnbland-wf** [comment]
Annnnnd it is back.

```
johnbland        96732   1.9 15.9  5171908 2662516   ??  S    12:48PM   5:14.55 /usr/local/Cellar/dart/1.23.0/libexec/bin/dart /usr/local/Cellar/dart/1.23.0/libexec/bin/snapshots/analysis_server.dart.snapshot --client-id=DanTup.dart-code --client-version=1.4.1 --instrumentation-log-file=dartcode.log
```

I'm cleaning up the log file then I'll post it.

**Comment by johnbland-wf** [comment]
A cleaner view of the memory:

![screen shot 2017-06-13 at 1 38 06 pm](https://user-images.githubusercontent.com/2642946/27098542-8f983c2a-503d-11e7-837e-998e198d624a.png)

**Comment by johnbland-wf** [comment]
Ok...the log is not revealing enough to paste the entire thing beyond this one bit to help replicate it.

- Nuke all packages: `rm -rf .pub .packages && find . -name packages | xargs rm -rf`
- `pub get --packages-dir` (in this case w/ packages)
- Memory jumps and the plugin is useless; `Analyzing...` shows in the footer and never goes away

If I try to run anything in Terminal, it fails until I quit VS Code.

**Comment by johnbland-wf** [comment]
```
1497378648559:Noti:{"event"::"server.status","params"::{"analysis"::{"isAnalyzing"::false}}}
1497378705215:Watch:/Users/johnbland/project:/Users/johnbland/project/pubspec.yaml:modify
...
1497378705215:Noti:{"event"::"server.status","params"::{"analysis"::{"isAnalyzing"::true}}}

```

**Comment by DanTup** [comment]
> the analyzer is kicked off in VS Code so I assume I'd need to add a setting to change it? 

Yep, there's a setting called `dart.analyzerObservatoryPort`. If you set that, then reload (you should be prompted to) then it'll be launched with observatory enabled for the analyzer. Then you can open your browser at `http://127.0.01:{port}/` to use Observatory. I'm not very familiar with it, but I expect there's a memory profile. I don't know how useful it'll be.

I think there's probably value in opening a case about this at dart-lang/sdk since the Analyzer team will probably have a much better idea of how to diagnose than me.

**Comment by DanTup** [comment]
There's also an option `dart.analyzerDiagnosticsPort` that lets you access a web app that may also give useful information (I can't remember exactly what it shows; not at a PC to check atm).

**Comment by DanTup** [comment]
@johnbland-wf I don't suppose you were able to do any further debugging of this? Can you still reliably repro it? And can you do so on a codebase that I could try and repro with (eg. is it public code, or can you repro with public code?).

**Comment by johnbland-wf** [comment]
@DanTup, I confirmed other users get the same issue on dart projects using this extension. We've learned to kill VS to recover and all is well.

Something is amiss and I don't have time to debug the extension issues at this point. If my time opens up, I'll revisit.

**Comment by DanTup** [comment]
@johnbland-wf Is the codebase you see this on public so I can try to repro? I'd very much like to fix this; it's very likely affecting other users too :(

@bwilkerson Looks like the Analyzer VM using the memory; does this seem like an unreasonable amount for it to use on a reasonable project? Is there any easier way to track it down without digging through Observatory (I'm happy to do that if I can get a repro on my machine).

I searched the sdk repo to see if there were any other reports of this and din't find much, though [this file](https://github.com/dart-lang/sdk/blob/ebd9440b18396124708cdb67bd82c1985e8a427a/tests/standalone/standalone.status) has a comment about a buildbot failing on Mac OS due to 256 file limit. Is it possible this is the cause?

**Comment by DanTup** [comment]
@johnbland-wf 

> We've learned to kill VS to recover and all is well

Btw, the "Reload Window" command *might* be a slightly easier way to get things working again - it should terminate the analyzer and reload the project in Code (which will start a new Analyzer). It'll still require the analyzer re-scan everything though.

**Comment by johnbland-wf** [comment]
No, the repo is not public, but it is quite large. You likely won't see this on small repos.

Reload Window does do what we need and that is to manage the analyzer instance running; hence the request for the extension to give us control of the services it is running.

**Comment by bwilkerson** [comment]
> Looks like the Analyzer VM using the memory; does this seem like an unreasonable amount for it to use on a reasonable project?

Kind of hard to say without knowing more about the size of the project being analyzed, the number of errors in the project, etc.

I'm assuming that this is a fairly recent build of the SDK. We've stopped keeping most of the data in memory that we used to cache, but there is certainly still some memory usage that is based on the size of the project.

> Is there any easier way to track it down without digging through Observatory (I'm happy to do that if I can get a repro on my machine).

Not that I'm aware of.

**Comment by johnbland-wf** [comment]
It is indeed a large project utilizing quite a large set dependencies and code from them. It is all Dart 1.24.2.

**Comment by johnbland-wf** [comment]
Just an FYI: numerous devs in the company use Code and are now realizing this is the source. It would be very helpful to provide a dart panel or basic actions to allow management of the analyzer service.

Here is your equivalent package for Atom and how it gives us control over the server as well as seeing basic stats.

![screen shot 2017-08-11 at 11 06 54 am](https://user-images.githubusercontent.com/2642946/29221677-46ba7cf4-7e85-11e7-8904-20643aa29195.png)

![screen shot 2017-08-11 at 11 07 54 am](https://user-images.githubusercontent.com/2642946/29221704-5b73b55c-7e85-11e7-98b7-cddc8f7a58b2.png)

**Comment by DanTup** [comment]
@johnbland-wf I'm slightly confused about what you're asking for. There's already a `Reload Window` command in Code which (if I understand correctly) does exactly what you're asking for. It's not a good fix; I'd rather we were able to eliminate the need for it, but it doesn't make sense to provide a duplicate command (nor would I like to make it look like it's an acceptable fox for this issue).

**Comment by johnbland-wf** [comment]
It is not a duplicate solution. It isn't even a solution, tbh, it is a workaround. Look at the screenshot. You can stop the analyzer, start, reanalyze, view status, etc.

I'm not sure how else to show how poorly "Reload Window" is as a solution; seeing as no one without knowledge of it being the fix would realize it. This plugin is making people stop using VS Code. A panel/command/otherwise to turn off or restart analyzer would be helpful.

**Comment by johnbland-wf** [comment]
And you do realize shortly after Reload Window it does the exact same thing again, right?

**Comment by Skylled** [comment]
I think the implication is that the Dart Analyzer shouldn't NEED to be restarted, and that the underlying issue should be addressed instead. Also that the same process of restarting the analyzer is performed by reloading the window.

Those extra options might make the idea more palatable though. All the same, the best option seems to be just throwing in a few new commands to the palette.

**Comment by bwilkerson** [comment]
> ... the underlying issue should be addressed instead.

I totally agree, and we'll try to allocate resources to look into it. Part of the goal, though, is to see if there's a short-term work around to help reduce user pain.

**Comment by DanTup** [comment]
@johnbland-wf Please understand that I really want to resolve your issues (and implement any useful features) but I can't implement something without fully understanding the problems.

> This plugin is making people stop using VS Code.

I think this is a little harsh. For many people, Dart Code works great (I finally got to use it myself recently for some work on the analysis server and with the exception of a couple of minor issues that I raised cases for, it worked very well, including running tests with `F5`). I personally know of several people that have moved *to* VS Code because of this plugin.

Please remember that this is an open source project - I don't work for Google, I don't get paid to work on this, my company doesn't benefit from this (we don't use Dart) and working on Dart Code is actually the main reason I don't actually get to write much Dart myself (because any time I have that I want to spend on something Dart-related is actually spent on this extension). I enjoy working on Dart Code because I learn lots of new stuff and I like to imagine it's useful to at least some people. Telling me things like my plugin is driving people away from Code is not the way to get me to implement the things you want; you need to be reasonable and you need to help me by answering my questions. I apologise if things that are obvious to you are not to me, I'm trying my best to help you while trying to avoid Dart Code descending into a random collection of features asked for by different people that I don't understand and have to maintain long-term. I care much more about delivering quality and being able to maintain and support Dart Code into the future than the number of features it has, so it's critical to me that I understand why every feature exists. I really don't feel I'm asking a lot in return for this free plugin.

> It is not a duplicate solution. It isn't even a solution, tbh, it is a workaround

I've already agreed with this several times, and so is `Restart Analysis Server` but at least this workaround is free and already there and not something I had to implement and maintain.

I don't want anyone having to reload their project (or restart the analysis server, or anything else like this). I gave it as an option because it sounds like you were asking for a feature that is functionally identical to it. If it's not, then I need to understand why (otherwise, I would presumably implement it incorrectly).

> Look at the screenshot. You can stop the analyzer, start, reanalyze, view status, etc.

I'm afraid I haven't written as much Dart as you so I don't entirely understand the need for each of these things. Can you explain the value of each? To my (very limited) eyes:

- `Restart Analysis Server` - This is functionally the same as `Reload Window` as far as I can tell and a hack/workaround for something going bad
- `Stop Analysis Server` - this will break the huge majority of functionality in Dart Code - I don't really understand why anyone would want to do this
- Status - We currently collect this info to include in error reports, but I'm not sure of the value of displaying it generally (we do show in the status bar when it's analyzing)

Can you provide more details about why/when you'd use each and for what?

> I'm not sure how else to show how poorly "Reload Window" is as a solution; seeing as no one without knowledge of it being the fix would realize it.

I'd argue the same for `Restart Analysis Server`. If Dart Code isn't working correctly how would the average user know that the fix is to `Restart Analysis Server`? Many users don't even know what the analysis server is and many of those that do don't know which parts of Dart Code it powers. IMO the analysis server is an implementation detail that the average user shouldn't need to worry about.

My main concern with implementing a specific command is that it will make it look like I'm ok with people having to restart the analysis server when they have issues, and I'm not. I would rather people came here and reported issues than quietly "fixed" the issues themselves, otherwise the real issue will never be fixed.

Also; if the issue is that the analysis server has crashed/not responding/whatevever - why should the user even need to run a command? Why can't Dart Code detect it and fix it silently in the background?

> A panel/command/otherwise to turn off or restart analyzer would be helpful.

Can you explain what you expect to happen if you "Turn off the analysis server"? The server is providing a significant chunk of the functionality in Dart Code (completion, errors, highlighting occurrences, go to definition, find references, etc.) - are you asking for a switch that would stop all of these from working? If so, which bits are left that you actually want (eg. that means you don't want to just disable the extension? Are you using it just for the debugger?)

> And you do realize shortly after Reload Window it does the exact same thing again, right?

I do not, because you hadn't told me. If that is the case, I would expect that restarting the analyzer will have the same issue, since it gets restarted when you reload the window? :(

I'm more than happy to try and resolve your issues; but you need to help me by answering my questions above. I have a growing list of requests and limited time so I'm not going to spend time on things that I don't believe will actually solves issues or deliver value. So far I don't understand the value of these features (or how they solve your problem).

**Comment by johnbland-wf** [comment]
You may feel it is harsh, but I'm letting you know how this is impacting us. We have a large swath of IntelliJ/WebStorm folks here and several switched to VS Code only to switch back after getting these errors. It seriously is a pain to have to deal with this numerous times throughout the day.

I moved to VS Code for this plugin. It just isn't working out great at this point so I've spent all of this time explaining the issue, providing debug logs, potential solutions we've used successfully before, etc in hopes of getting the plugin updated.

> Telling me things like my plugin is driving people away from Code is not the way to get me to implement the things you want; you need to be reasonable and you need to help me by answering my questions.

I've done that over and over again to only get `Reload Window` as the solution. This is impacting the entire OS...not just Code so it slows us down to have to close our dev environments (terminal, etc) then reopen them.

> I care much more about delivering quality and being able to maintain and support Dart Code into the future than the number of features it has, so it's critical to me that I understand why every feature exists. I really don't feel I'm asking a lot in return for this free plugin.

Great. I'm not arguing you do stuff that isn't useful. I just don't know how else to explain this situation to you, though. I've provided several options, debugging, etc to no avail. 

> I've already agreed with this several times, and so is Restart Analysis Server but at least this workaround is free and already there and not something I had to implement and maintain.
>
> I don't want anyone having to reload their project (or restart the analysis server, or anything else like this). I gave it as an option because it sounds like you were asking for a feature that is functionally identical to it. If it's not, then I need to understand why (otherwise, I would presumably implement it incorrectly).

I'm asking for a fix to this issue. Restarting the analyzer without having to reload my entire window was just a stopgap since reloading the window isn't discoverable at all; see people leaving Code not knowing how to "fix" the issue.

> Can you provide more details about why/when you'd use each and for what?

`Restart Analysis Server` (or `Reanalyze`) is an easy way to get the analyzer to restart for any reason not just extension issues. Sometimes it is outdated and didn't catch the branch changes or whatever happened and it says "File X does not exist" when it clearly does. Reload Window is expensive. It restarts every single extension/process/etc when only one thing is the problem: analyzer. `Reload Window` is killing a fly with a shotgun.

`Stop Analysis Server` allows Code to be an empty text editor without analyzer. I keep 2 windows open between the projects I am working on – sometimes simultaneously – and I don't care about analyzer data in the other. If I want to preserve memory but still see the files for reference as I shift things from one repo to another, being able to stop the analyzer could be quite useful in doing just that; especially seeing as the analyzer somehow runs into the GB of consumption per window at times.

`Status`, to start, would provide us the data necessary to debug this problem. Right now you've built a black box so devs won't have to see under the hood. At times, we need to and this is one. The Atom one shows me when the analyzer is working or idle (footer in Code shows that), but it also shows us the number of contexts, how many files were analyzed, and the task queue so we know if something is pending.

Here is the status section of the panel I mentioned working and clearly letting me know what's happening:

![gif](http://recordit.co/Sz95YFl4mr.gif)

All of that information helps us debug issues when analyzer trips up unrelated to this plugin as well as plugin issues.

Also note it shows what dart sdk is being used, seeing as multiple can exist on your machine at once.

> I'd argue the same for Restart Analysis Server. If Dart Code isn't working correctly how would the average user know that the fix is to Restart Analysis Server? Many users don't even know what the analysis server is and many of those that do don't know which parts of Dart Code it powers. IMO the analysis server is an implementation detail that the average user shouldn't need to worry about.

As noted above, reanalyzing your source can be highly useful. It is a targeted fix and most people absolutely should know exactly what Analyzer is if they're writing Dart code. Ignoring, as a dev, or hiding it as an "implementation detail" is handicapping yourself/your users for the sake of them not seeing under the hood. For new devs, maybe that helps. For experienced devs, it hurts.

Maybe we're different here, though. We built our own tools around the analyzer and make everyone learn about the analyzer plus fail our builds if it doesn't pass so everyone is intimately aware of what it does even if they're not familiar with the inner workings.

> My main concern with implementing a specific command is that it will make it look like I'm ok with people having to restart the analysis server when they have issues, and I'm not.

But the issue may not be the plugin. The issue could be in analyzer. Also, I'd say you're tacitly saying "restarting is ok" seeing as that is what `Reload Window` is doing.

> I would rather people came here and reported issues than quietly "fixed" the issues themselves, otherwise the real issue will never be fixed.

I did. I was told to reload the window. -_-

> I do not, because you hadn't told me. If that is the case, I would expect that restarting the analyzer will have the same issue, since it gets restarted when you reload the window? :(

From the thread, I thought that was understood. After opening Code and using it for a short period of time, we run out of files. Reloading the window is the same as opening Code so it happens again and again and again.

> I'm more than happy to try and resolve your issues; but you need to help me by answering my questions above. I have a growing list of requests and limited time so I'm not going to spend time on things that I don't believe will actually solves issues or deliver value. So far I don't understand the value of these features (or how they solve your problem).

As noted, I don't care about the features as much as fixing the issue but when I'm instructed to just reload the window I take that as a need for better features so I suggested some. Feel free not to implement any of the features but the issue absolutely needs some attention.

-----

Ultimately, I get this is a free extension. There is no need to impress upon me the virtues of open source/free...I'm well aware. I've spent 2-3 hours debugging/discussing this issue to help you fix this issue to help me continue using Code so I don't have to switch back to Atom – which it isn't all that bad, I just like Code.

All requests have been answered throughout this thread. You can see in the gif above there are 12,729 files analyzed. Something is amiss here. Maybe it is analyzer. Maybe it is the plugin. I can't say. My hope is you can.

**Comment by matanlurey** [comment]
@johnbland-wf - please don't be a [help vampire](http://www.skidmore.edu/~pdwyer/e/eoc/help_vampire.htm).

This is a free (as in beer) plugin, and nobody pays or sponsors @DanTup to work on this. If you have a team of dozens of developers affected, I'd consider either (a) sponsoring Danny (this is increasingly common in open source work), or (b) assigning someone to help Danny burn down issues you find relevant (i.e. contribute back).

**Comment by johnbland-wf** [comment]
LMBO @matanlurey.

**Comment by DanTup** [comment]
@johnbland-wf 

> You may feel it is harsh, but I'm letting you know how this is impacting us.

I totally understand that; but I'm really trying to help. I'm trying to have a dialogue so that I can completely understand the problems and try to provide the best solution. As developers I'm sure you find that you do better work when someone comes to you with a problem and lets you come up with solutions rather than telling you the solution they've come up with (especially if it's just the one that was provided somewhere else so they're used to it).

> I've done that over and over again to only get `Reload Window` as the solution

> I did. I was told to reload the window. -_-

>  but when I'm instructed to just reload the window

You have misinterpreted my comments. My first comment about using `Reload Window` was in response to you saying:

> We've learned to kill VS to recover and all is well.

It was intended as a short term fix (which is better/faster/less disruptive than "killing VS Code") while we figure out something better (which I hoped to be fixing the underlying issue or some other transparent fix that wouldn't involve the user having to mess around with internal services).

I suggested it again later (again, as a temporary workaround) because you kept asking for an option to `Restart the analysis server` and:

a) I believed (possibly incorrectly - see below) that `Reload Window` is basically identical to restarting the analysis server
b) I'd much rather unblock you today than have you wait for me to fix it and push a release out
c) I've tried to implement restarting the service before, to do it silently when the analysis server stopped responding and it was enough work that I decided to settle on just calling `Reload Window` ([for ex. here](https://github.com/Dart-Code/Dart-Code/blob/677d01ba4cee84f48bbef0231307d96a23cf1722/src/analysis/analyzer.ts#L82)).

Until your last comment I've had no idea why you believed that "restarting the analysis server" and "reload window" where not the same thing. Today you finally gave a hint about what this might be:

> Reload Window is expensive. It restarts every single extension/process/etc when only one thing is the problem: analyzer

For me, `Reload Window` takes around 2 seconds and puts everything back exactly how it was (same files open, etc.). It clears all state, aborts all in-flight requests and gives you a clean slate. Based on that, it seemed the same to me, it already existed and it requires no effort from me to use or maintain. I'm assuming based on this comment that reload window is not a quick operation for you? I've only worked on trivial codebases, I have no experience here.. How long does it take for you on your project? How many extensions do you have activated for your Dart projects?

As for the other requests:

> `Stop Analysis Server` allows Code to be an empty text editor without analyzer. I keep 2 windows open between the projects I am working on – sometimes simultaneously – and I don't care about analyzer data in the other.

> Maybe we're different here, though. We built our own tools around the analyzer and make everyone learn about the analyzer plus fail our builds if it doesn't pass so everyone is intimately aware of what it does even if they're not familiar with the inner workings.

You say everyone is intimately familiar with it, yet these comments ("I don't care about analyzer data" and "fail our builds if it doesn't pass") read to me like you might think it's providing only errors/warnings. I've tried to be careful to use the term `analysis server` in this thread because I don't want it confusing with the `dartanlyzer` CLI tool which just outputs errors/warnings. The analysis server is powering most of the major features of Dart Code (code completion, go-to-definition, finding symbols, highlighting occurrences, tooltips, etc). If I added a button to "Stop analysis server" it would disable *all of these things* making it hard for you to navigate around the code. Are you expecting this?

Also - if I implemented a new feature in Dart Code by parsing the Dart code myself which resulted in increased memory/CPU usage (eg. the analysis server didn't support it, or we were [trying out a new feature](https://github.com/Dart-Code/Dart-Code/issues/383) before adding it) would you like this button to also disable these features to save those resources too?

> (`Status`) but it also shows us the number of contexts, how many files were analyzed, and the task queue so we know if something is pending.

How often is this useful? What happens that makes you want to look at it? How long do you want to see it for? Does it need to be a real-time updating display (eg. vs just dumping it to the output window once)?

> Also note it shows what dart sdk is being used, seeing as multiple can exist on your machine at once.

Dart Code shows the SDK version in the status bar (bottom right), the same way that other languages (eg. TypeScript) do. Is this what you're asking for? If you set `dart.sdkPaths` you can even click on it to quickly switch SDKs (see #222).

> As noted above, reanalyzing your source can be highly useful. It is a targeted fix and most people absolutely should know exactly what Analyzer is if they're writing Dart code. Ignoring, as a dev, or hiding it as an "implementation detail" is handicapping yourself/your users for the sake of them not seeing under the hood. For new devs, maybe that helps. For experienced devs, it hurts.

It's not that I'm trying to hide it, but trying to simplify what's exposed to the users. I disagree that "most users" should need to know that code completion or highlighting occurrences is provided by a Dart program running in the background. I want complete Dart newbies to be able to install Dart Code and just start working with it without being confused about this "analysis service" that happens to be powering some of its features.

I need to find a balance. I want to support advanced users (there are Googlers that use it for Dart/Flutter/Fuchsia) but I don't want to over-complicate it for the average user. I'm not against implementing these features but I need to believe they're the best way to solve the problems and that they add more value than debt/complexity (currently I'm not so sure).


> But the issue may not be the plugin. The issue could be in analyzer. Also, I'd say you're tacitly saying "restarting is ok" seeing as that is what Reload Window is doing.

I'm not saying it's ok at all; it's always been offered as a dirty workaround while we discuss the options. If I thought it was the solution I would've closed this case a long time ago.

Anyway, let's try and get things moving rather having the same discussion. I have a Mac now, so I'll see if I can repro this by opening some large Dart repos. Can you provide answers on some of the new questions I've written above. I think it might be worth splitting this into separate issues so that they can discuss and implemented individually; for example:

- Too many open files errors
- High memory usage
- Analysis server diagnostic info (queue length, etc)
- Disable analysis server ("lightweight mode"?!)
- Restart analysis server (could this be automatic/silent in BG?)

Some of these make more sense to me than others, some are easy to do and others need more discussion/information. Having small specific cases will make it much easier to progress them or discuss whether there are better options for each one.

**Comment by johnbland-wf** [comment]
> As developers I'm sure you find that you do better work when someone comes to you with a problem and lets you come up with solutions rather than telling you the solution they've come up with (especially if it's just the one that was provided somewhere else so they're used to it).

If they are on my level or greater or have a good idea, yes I listen to their suggestions and consider it in my solution

------

I'll respond to the rest tomorrow.

**Comment by DanTup** [comment]
> If they are on my level or greater or have a good idea, yes I listen to their suggestions and consider it in my solution

Unless you're telling me that when someone "greater than your level" comes to you with a solution that you'd just implement it without understanding it, agreeing with it or asking questions about it then I believe we're agreeing.

I'm listening to and considering your suggestion - but currently I don't think it's the best solution. I'm not saying that it's not, I'm saying it doesn't seem like it to me based on the information I have. I want to be convinced by being given more information. I'm not just going to take your word for it because everyone believes their solution is best and they're biased towards their needs/preferences or what they're familiar with (I I had a pound for every time someone came to us saying "Can you implement (x feature that our old product had)?" and it turned out we could do much better I would have many pounds!). I'm trying to build a product that is best for most people, not just best for you.

You're asking me to spend a significant time implementing, testing and maintaining features. I'm asking you for a little time to answer my questions and discuss it so that I can be sure I'm implementing the best solutions. I don't think this an unfair trade in return for a free plugin (esp. if you're using it commercially to make money, it presumably has a non-zero value to you).

**Comment by johnbland-wf** [comment]
> Unless you're telling me that when someone "greater than your level" comes to you with a solution that you'd just implement it without understanding it, agreeing with it or asking questions about it then I believe we're agreeing.

Nope. That's not what I'm saying at all. The "agreeing with" part isn't always an option. What I said is I don't dismiss suggestions and openly accept them as a way to help me drill into a solution. You know...not believing I'm always right or the harborer of truth in solutions. #alwaysgrowing #alwayslearning

----

As to my suggestions, my goal is to fix the issues. I only suggested better solutions than "Reload Window" when digging into the issue wasn't clearly happening.

> I'm listening to and considering your suggestion - but currently I don't think it's the best solution. I'm not saying that it's not, I'm saying it doesn't seem like it to me based on the information I have.

That's fine. Your call.

> You're asking me to spend a significant time implementing, testing and maintaining features. I'm asking you for a little time to answer my questions and discuss it so that I can be sure I'm implementing the best solutions. I don't think this an unfair trade in return for a free plugin (esp. if you're using it commercially to make money, it presumably has a non-zero value to you).

I don't want it just for me. My suggestions would benefit all, but disagreeing with the features is fine. Let's just clean up the bug.

>  I'm not just going to take your word for it because everyone believes their solution is best and they're biased towards their needs/preferences or what they're familiar with (If I had a pound for every time someone came to us saying "Can you implement (x feature that our old product had)?" and it turned out we could do much better, I would have many pounds!). I'm trying to build a product that is best for most people, not just best for you.

No, I'm not. Again, no need to press "free plugin" here. If you aren't open to supporting bugs, just say that as opposed to guilt-tripping me for looking for it to be "usable".

**Comment by johnbland-wf** [comment]
And now that this ridiculous back and forth is done.

What have you seen in regards to this issue as possible approaches to resolve this issue?

**Comment by DanTup** [comment]
@johnbland-wf 

> What I said is I don't dismiss suggestions and openly accept them as a way to help me drill into a solution

I presume this comment is suggesting I'm not doing that; despite the fact I have repeatedly asked questions above to try and understand the problem so I can be sure we implement the best solution. I've said many times that I'm happy to implement your solution when I a) understand the problem fully and b) believe it's the best solution. If you won't help me achieve those, then the solutions will not be implemented. In the interests of progress, I'm not going to deal with any of the feature requests (restart/stop/status) here now - if you want them and are prepared to answer my questions about why you want them, please raise new issues and we'll have a discussion.

> I've answered every single question

I think we may have different understanding of the word "answered", but again, questions relate to features which can go into new issues if you want them so no more discussion on them here.

> It isn't too much to ask for a bug to be fixed.

Right, but the issue got rather derailed into feature requests that were never totally defined and you seemed to care more about.

As for the bug, it's not really that trivial - there are two issues (high memory, too many open files) and we don't know if they're bugs in Dart Code, the analysis server or VS Code. So far the issues are only occurring to you in a codebase I don't have access to and only after "a day or so". If you can help me reproduce it, I can investigate it faster. For example, if you can repro on a public repo and provide steps, maybe I can replicate it here.

> Again, no need to press "free plugin" here. If you aren't open to supporting bugs

I'm totally open to supporting bugs - I love making my software better; it actually upsets me when people have issues they don't report (something that's all too common). However, your attitude here has seemed to me like someone that has paid for support. I'm more than happy to support my software but in return I expect that people are polite, help me out with my questions and respect that I'm spending spare time on this (which I do not have a lot of).

> What have you seen in regards to this issue as possible approaches to resolve this issue?

So far, absolutely nothing. The time I hoped to spend on Dart Code this week has mostly been taken up trying to get more info on your feature requests and prior to that much of it was spent trying to understand why `Reload Window` is not as good a workaround as restarting the analysis server (a question I still don't have an answer to).

So, to bring this back on track, let's try and make it so I can reproduce these two issues (high memory usage, "too many open files") so I can investigate.

If it's not too much to ask, here's a couple of questions:

- What's the uptime for VS Code / the analysis server when you start hitting these (does it really only happen after a day, or can you cause it more quickly?)
- What's the structure of the open folder like? Is it a folder full of other dart projects (eg. with their own pubspecs)?
- Your screenshot shows approx 3k analyzed files - is that accurate for the project you're having issues on?
- How many files do you have open in VS Code (across all instances)?
- Can you repro on a public project (like the Dart SDK, for example)?

Also, while I'm unable to repro, maybe you could gather some info. There's [info here](https://askubuntu.com/a/181408) and how to possibly identify the process holding open files - this might help narrow things down.

**Comment by johnbland-wf** [comment]
@DanTup, I'm done w/ the back and forth. It is unproductive and I do not want to waste time explaining the confusions here.

> What's the uptime for VS Code / the analysis server when you start hitting these (does it really only happen after a day, or can you cause it more quickly?)

It varies. Sometimes it is quick (minutes) and other times it is longer (hour+), but that depends on usage.

> What's the structure of the open folder like? Is it a folder full of other dart projects (eg. with their own pubspecs)?

It is 1 project which, as of recently, has 2 pubspecs (a lib/ and an app/ folder). At the time of this issue it was 1 pubspec only.

> Your screenshot shows approx 3k analyzed files - is that accurate for the project you're having issues on?

Yup. That's the project.

> How many files do you have open in VS Code (across all instances)?

It varies. Usually just a few (the file + the test for it) and sometimes several when digging through files.

> Can you repro on a public project (like the Dart SDK, for example)?

No clue. I haven't tried it on any of them.

**Comment by johnbland-wf** [comment]
And re-read this one for steps to try to replicate the issue: https://github.com/Dart-Code/Dart-Code/issues/317#issuecomment-308213474.

**Comment by johnbland-wf** [comment]
Lastly, the comment above about changing settings seems to reference settings that are not available.

```
  // Whether SDK libraries should be marked as debuggable.
  "dart.debugSdkLibraries": false,

  // Whether external libraries should be marked as debuggable.
  "dart.debugExternalLibraries": false,

  // The location of the Dart SDK to use for analyzing and executing code. If blank, Dart Code will attempt to find it from the PATH environment variable.
  "dart.sdkPath": "",

  // The location of a folder that contains multiple Dart SDKs in sub-folders. When set, the version number in the status bar will be clickable to quickly switch between SDKs.
  "dart.sdkContainer": "",

  // The maximum length of a line of code. This is used by the document formatter.
  "dart.lineLength": 80,

  // Whether to insert argument placeholders during code completions.
  "dart.insertArgumentPlaceholders": true,

  // Whether to show TODOs in the Problems list.
  "dart.showTodos": true,

  // Offer to report analysis server exceptions when they occur.
  "dart.reportAnalyzerErrors": true,

  // Note: We only send a few very basic events and the platform and extension/Dart version numbers :-)
  "dart.allowAnalytics": true,

  // Checks whether you are using the latest version of the Dart SDK at startup.
  "dart.checkForSdkUpdates": true,

  // Additional args to pass to pub get and pub upgrade commands (eg. --packages-dir).
  "dart.pubAdditionalArgs": [],

  // Automatically runs `pub get` whenever pubspec.yaml is saved.
  "dart.runPubGetOnPubspecChanges": true,

  // Whether to show the names of linter rules in the problems panel to make it easier to `// ignore:`.
  "dart.showLintNames": false,

  // Whether to automatically send a 'hot reload' request during a Flutter debug session when saving files.
  "dart.flutterHotReloadOnSave": false
```

**Comment by DanTup** [comment]
> nm...they're private

Yeah; I keep thinking about "unhiding" them, because the editor giving squiggles is a bit misleading (it also bugs me that I see these warnings all the time, because I always have logging turned on). I'll put an issue in for the next release.

**Comment by johnbland-wf** [comment]
Is there something specific you want/need out of the server page or Observatory?

**Comment by johnbland-wf** [comment]
A googler noted the following:

> The cpu usage will likely be higher when used from VSCode. It's currently sending deltas on a per-keystroke basis; other clients (IntelliJ, Atom) batch things up until there's a pause in typing.

Potentially something to look at there?

**Comment by DanTup** [comment]
I presume there's a memory profiler in there, but I've never used it (yet!) so I'm not really sure I could give you instructions to extract anything useful without playing around with it myself first.

I had a go at trying to repro this with the Flutter repo; but so far the memory usage seems stable after the initial analysis.

> The cpu usage will likely be higher when used from VSCode. It's currently sending deltas on a per-keystroke basis; other clients (IntelliJ, Atom) batch things up until there's a pause in typing.

This is true; but nobody has ever mentioned high CPU to me so I've not investigated changing it. Are you having CPU issues in addition to the memory issues?

**Comment by johnbland-wf** [comment]
I haven't checked CPU usage regularly, but in checking it right now the Code Helper is holding 99-104% in Activity Monitor (Mac).

Idle:
<img width="411" alt="screen shot 2017-08-18 at 12 56 30 pm" src="https://user-images.githubusercontent.com/2642946/29471186-ae55acfa-8414-11e7-9b92-53226ab837fe.png">

**Comment by DanTup** [comment]
@johnbland-wf Is that just generally, or while you're typing?

Code Helper is used for a lot of things so it might be hard to pointpoint. There's some info in https://github.com/Microsoft/vscode/issues/11963 about tracking this down (`ps aux | grep <pid of process>` and get the full command line args). If it ends with `--type=extensionHost` it's an extension, but if not it could be a Code service (looks like some have had issues in the past). If it's the extensionHost, could you see if you can repro after restarting Code, and if so, disable Dart Code and restart to see if that fixes it. If it does, then please raise a new issue for investigation.

**Comment by johnbland-wf** [comment]
Yes, separate issue.

**Comment by DanTup** [comment]
@johnbland-wf Struggling to repro this.. Next time you see "too many open files", could you try the [commands here](https://superuser.com/questions/433746/is-there-a-fix-for-the-too-many-open-files-in-system-error-on-os-x-10-7-1/1180084#1180084) to see which process is holding the files open? (the full args will probably be helpful too).

**Comment by johnbland-wf** [comment]
Here ya go. I'm getting it now.

```
$ sudo lsof -n | cut -f1 -d' ' | uniq -c | sort | tail

 210 iconservi
 217 Chromium
 242 Dock
 258 com.apple
 369 Google
 525 Google
 532 Airmail
 670 Google
1468 Code\x20H
1776 Code\x20H
```

**Comment by DanTup** [comment]
@johnbland-wf Could you get the full command line args for those Code Helper processes (`ps aux | grep <pid of process>`)?

**Comment by johnbland-wf** [comment]
Those aren't PIDs.

**Comment by DanTup** [comment]
Oh yeah; crap! I think `sudo lsof -n | cut -f1,2,3,4 -d' ' | uniq -c | sort |  grep Code` will give the Code ones with their PIDs. If you take the ones with the high counts then run the command you should get the full args? (Not sure if that's best way, I'm a Windows guy)

**Comment by DanTup** [comment]
Oops, probably don't need that `,3,4`, I couldn't figure out why some didn't have pids!

**Comment by johnbland-wf** [comment]
```
johnbland        49265   0.0  6.9  4593000 1164220   ??  S     1:54PM   7:36.30 /Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper.app/Contents/MacOS/Code Helper /Applications/Visual Studio Code.app/Contents/Resources/app/out/bootstrap --type=watcherService

johnbland        49472   0.0  0.5  3379576  76768   ??  S     1:59PM   0:24.60 /Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper.app/Contents/MacOS/Code Helper /Applications/Visual Studio Code.app/Contents/Resources/app/out/bootstrap --type=watcherService
```

**Comment by johnbland-wf** [comment]
That's the last two:
```
$ sudo lsof -n | cut -f1,2 -d' ' | uniq -c | sort | grep Code

  35 Code\x20H 49274
  41 Code\x20H 49158
  42 Code\x20H 49278
  42 Code\x20H 49838
  44 Code\x20H 49470
  50 Code\x20H 49471
  52 Code\x20H 49189
  56 Code\x20H 49264
  75 Code\x20H 49167
  76 Code\x20H 49450
1298 Code\x20H 49472
1719 Code\x20H 49265
```

**Comment by Skylled** [comment]
Quick google search turned up this: https://github.com/Microsoft/vscode/issues/3998

**Comment by johnbland-wf** [comment]
That's high CPU. It could be related.

**Comment by DanTup** [comment]
@johnbland-wf Ok, that looks to me like a Code issue then :( If it was Dart Code I'd expect to see type=extensionHost on the end, and if it was the analysis server it would be in the dart process.

Would you mind raising it in the Code repo and posting the link here?

I might have an idea about the high RAM usage in the analysis server though; just looking...

**Comment by DanTup** [comment]
@Skylled Thanks for the link - this sounds like a possible culprit - it could explain the high CPU too!

**Comment by johnbland-wf** [comment]
It be interesting if that were the case. I only get this in VS Code on Dart projects and I leave VS Code open on my personal system for weeks at a time (JS and rails code).

**Comment by DanTup** [comment]
It is possible it's something specific in the Dart project causing it; but it doesn't look like Dart Code - our code runs in the extensionHost process and I think there's actually one or two places where Dart Code reads any files (since normally we're just getting contents from the buffers given to use by Code).

I'd post some info about your problem in that case - looks like it's been open a long time and not really progressed :(

**Comment by DanTup** [comment]
My theory about the RAM usage was incorrect, so back to drawing board on that one ;(

@johnbland-wf I haven't played around with it much yet, but possibly some useful info on memory usage from this:

- Set `"dart.analyzerObservatoryPort": 8081` in your settings
- Reload project to restart the analysis server
- Repro the problem/wait for memory usage to go high
- Load http://localhost:8081/ (or whatever port you used) in browser
- For the $main isolate (not the cache cleanup one), click Heap Snapshot
- Expand the largest branches of the tree

This should give you something like this:

<img width="1379" alt="screen shot 2017-08-18 at 20 53 35" src="https://user-images.githubusercontent.com/1078012/29475384-5e591640-8457-11e7-9642-3a30ffc4eb09.png">

This might indicate what's using up all the memory. I can't say how useful it'll be to me (I'm neither a Dart expert not familiar with the inner workings of the analysis server) but maybe something will jump out.

**Comment by johnbland-wf** [comment]
![screen shot 2017-08-18 at 3 00 39 pm](https://user-images.githubusercontent.com/2642946/29475585-089e3982-8426-11e7-9999-4b8eafe4823a.png)

This one is with a different grouping:
![screen shot 2017-08-18 at 3 01 51 pm](https://user-images.githubusercontent.com/2642946/29475633-388a2796-8426-11e7-883b-e2cfdbc9a5d4.png)

**Comment by DanTup** [comment]
@bwilkerson See comment above - I presume these FileSystemStates are contents of files being analysed. Is it expected that these would grow so large? I wondered if maybe I wasn't removing overlays when files are closed, but I just verified using the analysis server diagnostic app that the overlays are removed.

Any ideas? Or anything we can do to get more info? Would you rather move this over to an SDK issue?

**Comment by johnbland-wf** [comment]
BTW, Dart 1.25 dev has a MUCH better analysis server page. Parsing data there is much better.

**Comment by bwilkerson** [comment]
Yes, the FileSystemStates contain the contents of the files, as well as some other information to minimize the number of times we need to use I/O.

There's already an issue open: https://github.com/dart-lang/sdk/issues/30493.

**Comment by johnbland-wf** [comment]
I created it earlier, @bwilkerson.

**Comment by johnbland-wf** [comment]
Are you saying you keep an in-memory reference to every project file and every package we consume?

**Comment by johnbland-wf** [comment]
If I offended anyone or made them uncomfortable in any way, I sincerely apologize for doing so as that indeed was not my intent with this thread. It went off the rails, but my intentions are solely to make this extension work for us again as we enjoy VS Code + Dart Code.

I'll be back Monday to respond to any other feedback/info here. Thanks.

**Comment by DanTup** [comment]
@johnbland-wf 

I did write a long reply to your email, but (worried we'll just keep going in circles) decided I'd just summarise in response to you here and hopefully we can move on.

I'm not offended nor uncomfortable. I'm thick-skinned and happy to take criticism. I've learned a lot from discussing with people I disagree with (or just being told when I'm wrong) so any questions are not intended to start arguments, but rather for me to gain knowledge to correct/improve my understanding/opinions.

I apologise if I played a part in this going off the rails, I was frustrated by the lack of responses required to progress this. For example, my *very first* comment in this thread had a link to the command to find which process is holding files open - something you only finally responded to after another request yesterday, and it revealed that it's a VS Code watcher service - not Dart Code nor the analysis server.

I hope these issues get resolved. I think the memory usage issue is best handled in the SDK issue you raised and I hope if you haven't already, you'll raise an issue against Code for the "too many open files" issue (and post the link back here). I also hope this doesn't discourage you from raising issues against Dart Code in future - I'm still committed to supporting my users and remain happy to investigate issues that occur in Dart Code even if they might not be in Dart Code itself - with the provisos that if I cannot reproduce the issue I will need assistance and answers to questions, and that I won't just implement features as workarounds for bugs except as a last resort (where the bug cannot be tracked down or fixed).

Anyway; I hope this can be the end of this - it's been very distracting and I already have far more stuff to do than I'll ever fit into the free hours I have outside of work/family. There are no hard feelings; let's learn and move on.

**Comment by johnbland-wf** [comment]
Good deal.

Also, I never saw the link to the commands. That was my fault.

**Comment by DanTup** [comment]
I believe the conclusion was that there are two issues here, neither of which are Dart Code specific:

1. Analyzer memory usage grows too big over time (possibly caused by caching of file contents that are never evicted)
2. "Too many open files" errors, suspected to be caused by VS Code watcher service

As such, I'm closing this issue with the assumption there's nothing we can do in Dart Code to progress either (I'm trying to tidy up a little). If anyone thinks this isn't the case, please post details and we can re-open and investigate.
